### ISSUE TITLE
Use NLopt's version of `slsqp`

### ISSUE THREAD

**Issue by person142** [scipy/scipy#6689]
NLopt's version of `slsqp`

https://github.com/stevengj/nlopt/tree/master/slsqp

has some improvements over the original Fortran version--there are some bugfixes and the code is reentrant. I had thought we couldn't use it because NLopt was LGPL:

http://ab-initio.mit.edu/wiki/index.php/NLopt

but the Github repo appears to be MIT:

https://github.com/stevengj/nlopt/blob/master/COPYRIGHT

and the `slsqp` part retains its BSD license (inherited from SciPy) anyway:

https://github.com/stevengj/nlopt/blob/master/slsqp/COPYRIGHT

Might be worth replacing SciPy's version? (For concreteness: I volunteer to do it; just want to make sure there's not a reason we shouldn't.)

cc @stevengj

**Comment by stevengj** [comment]
See the [NLopt license](http://ab-initio.mit.edu/wiki/index.php/NLopt_License_and_Copyright).  If you link to the *whole* NLopt, you are bound by the GPL, because the Luksan code included in NLopt are GPL.  However, I tried to keep the GPL code segregated from the non-GPL code, so that if the GPL code is removed then the remainder reverts to a looser license.  In particular, my modifications to the SLSQP code remain under the BSD license, as you noticed.

**Comment by pv** [comment]
Not all of those fixes have been made in the modified fortran version in scipy, so just using the fixed version would avoid unnecessary work.

**Comment by mdhaber** [comment]
I thought I'd mention that there is another updated version of SLSQP:
https://github.com/jacobwilliams/slsqp

I'm not sure what the differences between that and the one in NLOPT are. 
@jacobwilliams would you be interested in your version of SLSQP replacing the one in SciPy?

**Comment by lucascolley** [comment]
does gh-22524 directly close this issue @ilayn ? We aren't exactly going to "use NLopt's version" right?

**Comment by ilayn** [comment]
Yes, I marked it as closing. NLOPT version is f2c translated Fortran code with changes that we also had in SciPy over the years.
